pkg/scripts/generate-variables.sh

```diff
@@ -1,67 +1,93 @@
 fi
 
-BACKEND_TF="${ROOT_FOLDER}/terraform-backend.tf"
-TFSTATE_LOCALS="${ROOT_FOLDER}/pkg/tfstate/tfstate-locals.tf"
-ASG_EC2_TAG_PATH="${ROOT_FOLDER}/pkg/modules/AWS/asg_ec2_tagging"
-
-# extract S3 bucket, dynamodb, tags, and region from locals.tf
-ENVIRONMENT_NAME=$(grep 'environment_name' ${CONFIG_FILE} | sed -nE 's/^.*"(.*)".*$/\1/p')
-REGION=$(grep 'region' ${CONFIG_FILE} | sed -nE 's/^.*"(.*)".*$/\1/p')
-
-# Get the AWS account ID
-AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
-
-# Generates the unique s3 bucket and key names for the deployment for keeping the terraform state
-S3_BUCKET="atlas-${ENVIRONMENT_NAME}-${REGION}-${AWS_ACCOUNT_ID}-tfstate"
-BUCKET_KEY="${ENVIRONMENT_NAME}-${AWS_ACCOUNT_ID}"
-
-# length of the bucket name should be less than 64 characters
-S3_BUCKET="${S3_BUCKET:0:63}"
-
-# Generates the unique dynamodb table names for the deployment lock ( convert all '-' to '_' )
-DYNAMODB_TABLE="tf_lock_${ENVIRONMENT_NAME//-/_}_${AWS_ACCOUNT_ID}"
-
-# Generate the terraform backend, where terraform store the state of the infrastructure
-echo "Generating the terraform backend definition file 'terraform.backend.tf'."
-sed 's/<REGION>/'${REGION}'/g'  "${SCRIPT_PATH}/../templates/terraform-backend.tf.tmpl" | \
-sed 's/<BUCKET_NAME>/'${S3_BUCKET}'/g' | \
-sed 's/<BUCKET_KEY>/'${BUCKET_KEY}'/g'  | \
-sed 's/<DYNAMODB_TABLE>/'${DYNAMODB_TABLE}'/g' \
-  > ${BACKEND_TF}
-
-# Generate the locals for terraform state
-echo "Generating the terraform state local file 'pkg/tfstate/tfstate-locals.tf'."
-sed 's/<REGION>/'${REGION}'/g'  ./pkg/templates/tfstate-locals.tf.tmpl | \
-sed 's/<BUCKET_NAME>/'${S3_BUCKET}'/g' | \
-sed 's/<BUCKET_KEY>/'${BUCKET_KEY}'/g'  | \
-sed 's/<DYNAMODB_TABLE>/'${DYNAMODB_TABLE}'/g' \
-  > ${TFSTATE_LOCALS}
-
-
-
-# fetch the config files from root
-cp -fr "${SCRIPT_PATH}/../../variables.tf" "${SCRIPT_PATH}/../tfstate"
-cp -fr "${CONFIG_FILE}" "${SCRIPT_PATH}/../tfstate"
-
-# copy variable files for tagging module
-cp "${SCRIPT_PATH}/../../${CONFIG_FILE}" "${ASG_EC2_TAG_PATH}"
-cp "${SCRIPT_PATH}/../../variables.tf" "${ASG_EC2_TAG_PATH}"
-cp "${SCRIPT_PATH}/../tfstate/tfstate-locals.tf" "${ASG_EC2_TAG_PATH}"
\ No newline at end of file
+show_help() {
+    echo "The terraform config filename for infrastructure is missing."
+    echo
+    echo "Usage: generate-variables.sh <config_file> [<path_to_root>]"
+    exit 1
+}
+
+set_variables() {
+  echo ${CONFIG_FILE}
+  # extract S3 bucket, dynamodb, tags, and region from locals.tf
+  ENVIRONMENT_NAME=$(grep 'environment_name' ${CONFIG_FILE} | sed -nE 's/^.*"(.*)".*$/\1/p')
+  REGION=$(grep 'region' ${CONFIG_FILE} | sed -nE 's/^.*"(.*)".*$/\1/p')
+
+  # Get the AWS account ID
+  AWS_ACCOUNT_ID=$(aws sts get-caller-identity --query Account --output text)
+
+  # Generates the unique s3 bucket and key names for the deployment for keeping the terraform state
+  S3_BUCKET="atlas-${ENVIRONMENT_NAME}-${REGION}-${AWS_ACCOUNT_ID}-tfstate"
```

Comment:
Are we creating s3 bucket per each environment? I vaguely remember that we create one bucket, and put different environment deployments into sub-folders, i.e. bucket key.

Reply:
That was for ages ago, as far as we handle multiple products in one environment now, we have decided to have a separated S3 for each environment to be more manageable and we can cleanup all resources including s3 independently.

Reply:
I see. yup, make sense. Thanks for the clarification! 

In this case, if user used all 25 characters as environment name, S3 bucket name will be trimmed to 64 chars in the following line. It then looks like `atlas-environmentname-ap-southeast-1-123456789012-tfst`. The last bit `tfst` looks a bit ugly to me... Since it is only 3-character difference, is it possible to limit the environment name to 22 chars long?

Reply:
I totally understand your point. However no one look at this names in normal circumstances. We can even drop the postfix at all. This case is the same for cluster name. We may drop part or all part of cluster name postfix which is totally okay. The most important thing is we still make the cluster name and s3 bucket name unique. For that reason I will leave the length of environment name to the maximum available length that keep the s3 and cluster names unique, in our case 25 characters. This gives the maximum flexibility to user to choose environment name.

Reply:
Where does 25 characters come from when you first decide it? cos 3-character difference doesn't seem significant to me. User wont lose too much flexibility. 
 
And you're absolutely right when looking terraform infra solely, probably resource name isn't that important. However, I'm more thinking from a user perspective. I wouldn't be surprised that user will have other workloads in the same aws account. Keeping the name consistent will provide user some pattern to easily opt in or filter out Atlassian deployment artifacts, maybe used for some reporting purpose or scheduled jobs to scan certain resources. `xxx-tfstate`, `xxx-tfstat` or `xxx.tfst` doesn't provide such consistency/pattern. And same apply for other aws resources we created. 

Another option I can think of is using `atl` instead of `atlas` as prefix, in this case, there are still 24 chars avail to user.

Reply:
Good point. What I did, considered all places we use environment name and picked the one with most constrains to define the maximum length for this item. 
Saying that, it comes from cluster naming - `atlassian-dc-<env>-cluster`.
As cluster name can be max 38 characters so considering sacrifice the postfix `-cluster`, we can have maximum 25 characters for environment name.

Reply:
I edited postfix and prefix for VPC, cluster, S3, and EFS names to make them consistent and also avoid postfix sacrifice and at the same time keep 25 length for environment name.